File: 2024/March/3-2/SortedArraySquares.js
const sortedSquares = function(nums) {
    if(nums.length === 1) {
        return [nums[0] * nums[0]];
    }

    var posI = 0;
    const sorted = new Array(nums.length);
    var sortedIndex = 0;

    while(posI < nums.length && nums[posI] < 0) {
        posI++;
    }
    var negI = posI - 1;

    while(negI >= 0 && posI < nums.length) {
        var posSqr = nums[posI] * nums[posI];
        var negSqr = nums[negI] * nums[negI];

        if(posSqr < negSqr) {
            sorted[sortedIndex] = posSqr;
            posI++;
            sortedIndex++;
        }
        else {
            sorted[sortedIndex] = negSqr;
            negI--;
            sortedIndex++;
        }
    }

    while(negI >= 0) {
        sorted[sortedIndex] = nums[negI] * nums[negI];
        negI--;
        sortedIndex++;
    }

    while(posI < nums.length) {
        sorted[sortedIndex] = nums[posI] * nums[posI];
        posI++;
        sortedIndex++;
    }

    return sorted;
};

const nums = [-7,-3,2,3,11];
console.log(sortedSquares(nums));

File: 2024/March/3-2/SortedArraySquares.java
public class SortedArraySquares {
    public static void main(String[] args) {
        int[] nums = {-7, -3, 2, 3, 11};
        int[] sol = new Solution().sortedSquares(nums);

        for(int i: sol) {
            System.out.print(i + " ");
        }
    }
}

class Solution {
    public int[] sortedSquares(int[] nums) {
        /* Base case */
        if(nums.length == 1) {  
            return new int[] { nums[0] * nums[0]};
        }

        /* Variables */
        int positiveIndex = 0;
        int negativeIndex;
        int[] sorted = new int[nums.length];
        int sortedIndex = 0;

        /* Find the positive/negative point */
        while(positiveIndex < nums.length && nums[positiveIndex] < 0) { 
            positiveIndex++;
        }
        negativeIndex = positiveIndex - 1;

        /* Insert into solution array in order */
        int positiveSquare;
        int negativeSquare;
        while(negativeIndex >= 0 && positiveIndex < nums.length) {  // Move pointers outwards until bounds of num
            positiveSquare = nums[positiveIndex] * nums[positiveIndex];
            negativeSquare = nums[negativeIndex] * nums[negativeIndex];

            if(positiveSquare < negativeSquare) {
                sorted[sortedIndex++] = positiveSquare;
                positiveIndex++;
            } else {
                sorted[sortedIndex++] = negativeSquare;
                negativeIndex--;
            }
        }

        /* Complete nums traversal */
        while(negativeIndex >= 0) {
            sorted[sortedIndex++] = nums[negativeIndex] * nums[negativeIndex];
            negativeIndex--;
        }

        while(positiveIndex < nums.length) {
            sorted[sortedIndex++] = nums[positiveIndex] * nums[positiveIndex];
            positiveIndex++;
        }

        return sorted;
    }
}